proxy/envoy/config.go

```diff
@@ -124,83 +112,126 @@ func Generate(instances []*model.ServiceInstance, services []*model.Service, mes
 	}, nil
 }
 
-func buildIngressListeners(instances []*model.ServiceInstance) ([]Listener, []Cluster) {
-	// project to port values
-	ports := make(map[int]map[model.Protocol]bool)
-	for _, instance := range instances {
-		protocols, ok := ports[instance.Endpoint.Port.Port]
-		if !ok {
-			protocols = make(map[model.Protocol]bool)
-			ports[instance.Endpoint.Port.Port] = protocols
+func buildListeners(instances []*model.ServiceInstance, services []*model.Service) ([]Listener, []Cluster) {
+	clusters := buildClusters(services)
+	listeners := make([]Listener, 0)
+
+	// group by port values
+	type listener struct {
+		instances map[model.Protocol]bool
+		// services map service, port for a given port, protocol pair
+		services map[model.Protocol][]*model.Service
+	}
+
+	ports := make(map[int]*listener, 0)
+	ensure := func(port model.Port) {
+		if _, ok := ports[port.Port]; !ok {
+			ports[port.Port] = &listener{
+				instances: make(map[model.Protocol]bool),
+				services:  make(map[model.Protocol][]*model.Service),
+			}
 		}
-		protocols[instance.Endpoint.Port.Protocol] = true
 	}
 
-	listeners := make([]Listener, 0)
-	clusters := make([]Cluster, 0)
-	for port, protocols := range ports {
-		cluster := Cluster{
-			Name:             fmt.Sprintf("%s%d", IngressClusterPrefix, port),
-			Type:             "static",
-			ConnectTimeoutMs: 1000,
-			LbType:           LbTypeRoundRobin,
-			Hosts:            []Host{{URL: fmt.Sprintf("tcp://%s:%d", "127.0.0.1", port)}},
+	for _, instance := range instances {
+		port := instance.Endpoint.Port
+		ensure(port)
+		ports[port.Port].instances[port.Protocol] = true
+	}
+
+	for _, svc := range services {
+		for _, port := range svc.Ports {
+			ensure(port)
+			ports[port.Port].services[port.Protocol] = append(ports[port.Port].services[port.Protocol], &model.Service{
+				Name:      svc.Name,
+				Namespace: svc.Namespace,
+				Ports:     []model.Port{port},
+			})
 		}
+	}
 
+	for port, lst := range ports {
 		listener := Listener{
 			Port:       port,
 			BindToPort: false,
 		}
 
+		ingressClusterName := fmt.Sprintf("%s%d", IngressClusterPrefix, port)
```

Comment:
ingressCluster seems wrong terminology to indicate a single app instance co-located with Envoy. May be "localAppInstance", where instance clearly indicates that its a singleton, local (to imply that its forwarding traffic to the app in same pod).

Also, won't this break in per-node Envoy deployment?